### PR TITLE
ifdef pthread_getname_np as musl doesn't have it

### PR DESCRIPTION
**Pull request by z3ntu**

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
I guess I'd start another debate if I mentioned adding a
```c++
void pthread_getname_np(...) {}
```
overload

**Comment by z3ntu** [comment]
I don't particularly care about how it's solved; this is my attempt of solving it. If you have a better idea, feel free to do that. :) (you can push new commits to this PR so you don't have to create a new PR)

**Comment by z3ntu** [comment]
Well it's broken, who would've expected... Will fix it

**Comment by AlanGriffiths** [comment]
I'm going to rework this.(I'll try pushing to this branch).

I agree with @wmww that it looks odd to check for a symbol only used by tests in the root CMakeLists.txt. But my bigger concern is that we shouldn't, for example, be frigging `ThreadedDispatcherTest.sets_thread_names_appropriately` so that it passes, we should be disabling the test as it does not test what it claims.

**Comment by AlanGriffiths** [comment]
> I actually can't move the check because if the check is after the -Werror addition, it will fail because "unused variable" warnings that exist in the check code will error out and will mark the variable not found.

Where do you see this behavior? (I don't see it on Ubuntu 18.04)

**Comment by z3ntu** [comment]
On Arch Linux, putting the following into `CMakeLists.txt` 
```cmake
include(CheckSymbolExists)
list(APPEND CMAKE_REQUIRED_DEFINITIONS -D_GNU_SOURCE)
list(APPEND CMAKE_REQUIRED_LIBRARIES "-lpthread")

# Before -Werror
check_symbol_exists(pthread_getname_np pthread.h HAVE_PTHREAD_GETNAME_NP)

# Set CFLAGS
set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -pthread -g -Werror -Wall -pedantic -Wextra -fPIC")

# After -Werror
check_symbol_exists(pthread_setname_np pthread.h HAVE_PTHREAD_SETNAME_NP)
```
and running `cmake .` in the same directory results in (other stuff excluded)
```
-- Looking for pthread_getname_np
-- Looking for pthread_getname_np - found
-- Looking for pthread_setname_np
-- Looking for pthread_setname_np - not found
```

**Comment by AlanGriffiths** [comment]
@wmww WDYT?

(I still think using the compiler to detect the function is better than using CMake and the pre-processor, but I know I'm in a minority.)

**Comment by Conan-Kudo** [comment]
@z3ntu Why do you know why doesn't musl have this function? I don't know of a reason why it shouldn't have it...

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -105,6 +105,13 @@ endif()
 set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fno-lto")
 set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-lto")
 
+include(CheckSymbolExists)
+list(APPEND CMAKE_REQUIRED_DEFINITIONS -D_GNU_SOURCE)
+list(APPEND CMAKE_REQUIRED_LIBRARIES "-lpthread")
+string(REPLACE " -Werror " " " CMAKE_C_FLAGS ${CMAKE_C_FLAGS})    # This flag breaks check_symbol_exists()
```
Comment (wmww) [inline tests/CMakeLists.txt]: Will this have any unintended consequences, and does `check_cxx_symbol_exists()` work without this line?

**Comment by AlanGriffiths** [comment]
> @z3ntu Why do you know why doesn't musl have this function? I don't know of a reason why it shouldn't have it...

Agreed, I also wondered if it would be simple to implement a replacement.

**Comment by z3ntu** [comment]
No, I don't know. 

> musl is a “libc”, an implementation of the standard library functionality described in the ISO C and POSIX standards, plus common extensions

Probably this function isn't common enough?

**Comment by Conan-Kudo** [comment]
@z3ntu In general, when musl doesn't implement something, instead of making applications work around libc, you should ask the musl people _first_ if they can implement it. API compatibility is an explicit goal of the project, after all...

**Pull request merged**